### PR TITLE
feat(consensus): additional sanity checks for the size of proposed blocks

### PR DESCRIPTION
**Pull request by cason**
Hardens tests regarding the size of proposed blocks, namely:

- The byte size of a proposal block `Part` should be constant (`== types.BlockPartSizeBytes`), except for the last part of a `PartSet`  (`<= types.BlockPartSizeBytes`)
-  A valid `Proposal` should not enclose a `PartSet` enabling the building of a `ProposalBlock` with size larger than the configured `ConsensusParams.Block.MaxBytes`. Notice that building a `ProposalBlock` larger than the allowed would fail in any case, but the proposed changes also invalidate the associated `Proposal`.

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: types/part_set.go
Comment (sergio-mena) [inline types/part_set.go:33]: I know this is not strictly necessary, but we can turn this `fmt.Error` into a proper error as we're doing below?

File: types/part_set.go
Comment (sergio-mena) [inline types/part_set.go:36]: Shouldn't this condition be using the actual Index of the part, rather than the index of the proof?

File: types/part_set.go
Comment (sergio-mena) [inline types/part_set.go:37]: We're still using `part.Proof.Total` here, are 100% sure this total will always be in sync with the total number of parts?

Comment (cason) [reply types/part_set.go:37]: Yes, this is a good point. But we don't have the `Total` field in each `Part`, we only have it on the `PartSet`.

Comment (cason) [reply types/part_set.go:37]: The point here is that wrong values for `Proof.Total` and `Proof.Index` should make the Merkle-tree verification to fail. But I am not sure whether we test this at this point. Notice that the `Part` tests have nothing on the arrays used to fill the `Part` and the `Part.Proof` bytes.

File: consensus/state.go
Comment (cason) [reply consensus/state.go:1985]: Wow, this really a bug. The code should be the same as in the following method, namely the test is whether `maxBytes == -1`, meaning that it is _disabled_ .

Comment (sergio-mena) [reply consensus/state.go:1985]: Thanks @lasarojc for spotting this!

File: types/part_set.go
Comment (sergio-mena) [reply types/part_set.go:37]: Is there a possibility that `Part` is used without populating `part.Proof`? (I guess no, but double-checking).

File: consensus/state_test.go
Comment (sergio-mena) [inline consensus/state_test.go:317]:
```suggestion
			maxBlockParts := (maxBytes-1) / int64(types.BlockPartSizeBytes) + 1
			numBlockParts := int64(propBlockParts.Total())
```

How about this instead?

Comment (cason) [reply consensus/state_test.go:317]: I prefer using a different method, even if it is less efficient/smart, when testing.

File: types/part_set.go
Comment (cason) [reply types/part_set.go:37]: It should not be possible, as the next check (line 40) would fail. This check, however, is also a valid check: integer fields are >= 0, hashes have the right length.

File: consensus/state_test.go
Comment (sergio-mena) [reply consensus/state_test.go]: Good point

File: types/part_set.go
Comment (cason) [reply types/part_set.go:37]: So, status regarding this:
- we can trust `Proof.Index`, as change it causes the proof verification to fail
- we can't trust `Proof.Total`, as it is irrelevant for the proof verification provided that `Proof.Index < Proof.Total`.

Comment (cason) [reply types/part_set.go:37]: Ok, now `Proof.Total` should be correct, otherwise the `Part` is not added. Lets check if something breaks due to this additional check.

Comment (sergio-mena) [reply types/part_set.go:37]: Excellent!

**Review by sergio-mena — APPROVED** [review]
I was convinced I had already approved this (sorry for delay)

**Comment by cason** [comment]
> Is this going to be backported? If so, to which version(s)?

I would say all releases. Why?

**Comment by thanethomson** [comment]
> I would say all releases. Why?

There are currently no backport labels on the PR.

Let's get this over the finish line please.

**Comment by cason** [comment]
> There are currently no backport labels on the PR.

Sorry, my bad. @sergio-mena, can you confirm we are backporting this to all releases?

**Comment by sergio-mena** [comment]
As this comes from our work in a security advisory, I'd say we should backport it to all releases

**Comment by thanethomson** [comment]
What's blocking the merge here @cason?

**Comment by andynog** [comment]
Moving this to `In Progress` since looks like more input is needed here. When it's `Ready for review` again please change the status.

File: consensus/state_test.go
Comment (sergio-mena) [reply consensus/state_test.go:314]: IIRC, @cason wanted to have it calculated differently in the test code and in the prod code, and I agreed with him

**Comment by cason** [comment]
So, the main question that worries me regarding this PR is the following: is it a good idea to enforce block parts to have exactly `types.BlockPartSizeBytes`?

I can think on reasons for a proposer to opt for smaller parts, for performance or network dissemination reasons. And this should not be considered a bad behavior. In any case we already reject, at the reactor level, parts that are bigger than that.

The main reason we had for enforcing a fixed size for inner block parts is to infer the size of a proposed `Block` from the `BlockID` included in the associated `Proposal`, as its size would be at least its `(BlockID.PartSetHeader.Total- 1) x types.BlockPartSizeBytes`. The point is that, while less precise, we still get a decent upper bound for the size of a proposed block if we don't enforce a fixed size for inner block parts. We can still use the same logic to detect proposals for blocks that cannot be accepted in any way, despite the size of the block parts.

Notice that in any case, as reported in some comments, this logic to enforce all but the last block parts to have a fixed size is kind of hacky, because we use indexes and total number of parts from the Merkle proof etc. We could make this less hacky if we moved this verification from `Part.ValidateBasic()` to the consensus logic.

The last point, also for the same reasoning, is that types `Part` and `PartSet` are public, exported, and we don't really know the impact on enforcing this requirement for them. As the changes in our test units indicate, there is the possibility and maybe it is even common to build part set using different part sizes.

I would like some feedback on this before rendering this PR ready to merge.

**Comment by melekes** [comment]
> The main reason we had for enforcing a fixed size for inner block parts is to infer the size of a proposed Block from the BlockID included in the associated Proposal,

Is this a reason for a fixed size or capping maximum size? I.e. why do we need a more precise value for a proposal block size?

**Comment by cason** [comment]
> Is this a reason for a fixed size or capping maximum size? I.e. why do we need a more precise value for a proposal block size?

For fixed, not a maximum. I had the same confusion while writing. The introduced logic rejects the `Proposal` for a block when

`[num of parts - 1] x [part size] > [maximum configured block size]`

If someone uses smaller part sizes, we might reject proposals for blocks whose size does not exceed the configured maximum.

**Comment by melekes** [comment]
> > Is this a reason for a fixed size or capping maximum size? I.e. why do we need a more precise value for a proposal block size?
> 
> For fixed, not a maximum. I had the same confusion while writing. The introduced logic rejects the `Proposal` for a block when
> 
> `[num of parts - 1] x [part size] > [maximum configured block size]`
> 
> If someone uses smaller part sizes, we might reject proposals for blocks whose size does not exceed the configured maximum.

In that case enforcing size makes sense to me 👍

**Comment by sergio-mena** [comment]
@cason Did we see a valid use case being used on the filed where the proposer is using block parts of sizes different from the default? (except the last block part, obviously)

**Comment by faddat** [comment]
I think this PR is a great idea.

**Review by faddat — APPROVED** [review]
I really like this approach as a way to further defend against storms.

**Review by melekes — APPROVED** [review]

**Comment by cason** [comment]
Ok, lets land this on `main` like it is. I will propose another way to implement the same without forcing generic `PartSet`s to adhere to the sizes required by consensus.

**Comment by cason** [comment]
A potential use-case for smaller block parts: #1776

**Pull request merged**
